cdc_rsync_server/cdc_rsync_server.cc

```diff
@@ -126,11 +164,78 @@ class PatchTask : public Task {
     status_ = path::SetFileTime(target_filepath_, file_.client_modified_time);
   }
 
+  const std::string base_filepath_;
+  const std::string target_filepath_;
+  const ChangedFileInfo file_;
+  CdcInterface* const cdc_;
+  const bool need_intermediate_file_ = false;
+  const std::string patched_filepath_;
+
+  FILE* patched_fp_ = nullptr;
+  bool is_executable_ = false;
+  absl::Status status_;
+
+  // This task is queued twice, once to patch and once to close and finalize the
+  // patched file.
+  enum class State { kPatching, kFinalizing, kDone };
+  State state_ = State::kPatching;
+};
+
+// Background task that closes a file and sets the mtime and perms. This is done
+// in the background since closing a file might block for a long time.
+class FinalizeCopiedFileTask : public Task {
+ public:
+  // Finalize |file| with given path |filepath|. |status| is the status from
+  // writing the file. On error, the file is only closed.
+  FinalizeCopiedFileTask(FILE* fp, FileInfo file, std::string filepath,
+                         bool is_executable, absl::Status status)
+      : fp_(fp),
+        file_(std::move(file)),
+        filepath_(std::move(filepath)),
+        is_executable_(is_executable),
+        status_(status) {}
+  virtual ~FinalizeCopiedFileTask() = default;
```

Comment:
Why is this line needed? I would expect `Task` to have a virtual destructor and if not it should probably have one. And then there is no need to redeclare if virtual here.

Reply:
Not strictly necessary, but I would consider it good practice. My main reason is, if you don't declare a destructor, it'll be inlined in the .h file (in this case the class is defined and declared in .cc, so irrelevant here). If you forward declare a type, and you have a std::unique_ptr of that type, this will fail since it won't know in the .h file how to destroy it. If you define the constructor as default in the .cc file, it will know how to destroy the pointer.